Documentation/DeveloperCorner/ConfigurationAPI.rst
..  include:: /Includes.rst.txt

..  _configuration-api:

=================
Configuration API
=================

In order to access the :ref:`extension configuration <extension-configuration>`,
a slim PHP API exists. Each configuration option is accessible by
an appropriate class method.

..  php:namespace:: EliasHaeussler\Typo3Solver\Configuration

..  php:class:: Configuration

    API to access all available extension configuration options.

    ..  php:method:: getAPIKey()

        Get the configured :ref:`API key <extconf-api-key>`.

        :returntype: string|null

    ..  php:method:: getModel()

        Get the configured :ref:`AI model <extconf-attributes-model>`.

        :returntype: string

    ..  php:method:: getMaxTokens()

        Get the configured :ref:`maximum number of tokens <extconf-attributes-maxTokens>`.

        :returntype: int

    ..  php:method:: getTemperature()

        Get the configured :ref:`temperature <extconf-attributes-temperature>`.

        :returntype: float

    ..  php:method:: getNumberOfCompletions()

        Get the configured :ref:`number of completions <extconf-attributes-numberOfCompletions>`.

        :returntype: int

    ..  php:method:: getCacheLifetime()

        Get the configured :ref:`cache lifetime <extconf-cache-lifetime>`.

        :returntype: int

    ..  php:method:: isCacheEnabled()

        Check whether solution caching is currently enabled. The cache is
        enabled if the configured :ref:`cache lifetime <extconf-cache-lifetime>`
        is greater than :php:`0`.

        :returntype: bool

    ..  php:method:: getProvider()

        Get the configured :ref:`solution provider <extconf-provider>`.

        :returntype: :php:interface:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Provider\\SolutionProvider`

    ..  php:method:: getPrompt()

        Get the configured :ref:`prompt <extconf-prompt>`.

        :returntype: :php:interface:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Prompt\\Prompt`

    ..  php:method:: getIgnoredCodes()

        Get the configured :ref:`list of exception codes to ignore <extconf-ignoredCodes>`.

        :returntype: array

..  seealso::

    View the sources on GitHub:

    -   `Configuration <https://github.com/eliashaeussler/typo3-solver/blob/main/Classes/Configuration/Configuration.php>`__


Documentation/Configuration/ExtensionConfiguration.rst
..  include:: /Includes.rst.txt

..  _extension-configuration:

=======================
Extension configuration
=======================

The extension currently provides the following configuration options:

..  _extconf-provider:

..  confval:: provider
    :type: string (FQCN)
    :Default: :php:class:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Provider\\OpenAISolutionProvider`

    Default solution provider to be used for problem solving.

    ..  note::

        Custom solutions providers must implement
        :php:interface:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Provider\\SolutionProvider`.

..  _extconf-prompt:

..  confval:: prompt
    :name: extconf-prompt
    :type: string (FQCN)
    :Default: :php:class:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Prompt\\DefaultPrompt`

    Default prompt to be used for problem solving.

    ..  note::

        Custom prompts must implement
        :php:interface:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Prompt\\Prompt`.

..  _extconf-ignoredCodes:

..  confval:: ignoredCodes
    :type: string (comma-separated list)

    Comma-separated list of exception codes to ignore during problem solving.
    Once a raised exception matches with the configured exception codes, the
    solution provider won't provide a solution and is simply ignored.

    ..  note::

        This setting only applies to the default
        :php:class:`EliasHaeussler\\Typo3Solver\\ProblemSolving\\Solution\\Provider\\OpenAISolutionProvider`.

..  _extconf-api-key:

..  confval:: api.key
    :type: string

    :ref:`API key <api-key>` for OpenAI requests.

    ..  attention::

        It is essential to configure the API key. Otherwise, requests to OpenAI
        won't be possible.

..  _extconf-attributes-model:

..  confval:: attributes.model
    :type: string
    :Default: `gpt-4o-mini`

    `OpenAI model <https://platform.openai.com/docs/models>`__ to use (see
    :ref:`List available models <solver-list-models>` to show a list of available
    models).

..  _extconf-attributes-maxTokens:

..  confval:: attributes.maxTokens
    :type: integer
    :Default: `300`

    `Maximum number of tokens <https://platform.openai.com/docs/api-reference/chat/create#chat-create-max_completion_tokens>`__
    to use per request to OpenAI.

..  _extconf-attributes-temperature:

..  confval:: attributes.temperature
    :type: float
    :Default: `0.5`

    `Temperature <https://platform.openai.com/docs/api-reference/chat/create#chat/create-temperature>`__
    to use for OpenAI completion requests (must be a value between `0` and `1`).

..  _extconf-attributes-numberOfCompletions:

..  confval:: attributes.numberOfCompletions
    :type: integer
    :Default: `1`

    `Number of completions <https://platform.openai.com/docs/api-reference/chat/create#chat/create-n>`__
    to generate for each problem.

..  _extconf-cache-lifetime:

..  confval:: cache.lifetime
    :type: integer
    :Default: `86400` *(= 1 day)*

    Lifetime in seconds of the solutions cache.

    ..  tip::

        Use `0` to disable caching.


Classes/Configuration/Configuration.php
<?php

declare(strict_types=1);

/*
 * This file is part of the TYPO3 CMS extension "solver".
 *
 * Copyright (C) 2023-2025 Elias Häußler <elias@haeussler.dev>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

namespace EliasHaeussler\Typo3Solver\Configuration;

use EliasHaeussler\Typo3Solver\ProblemSolving;
use TYPO3\CMS\Core;

/**
 * Configuration.
 *
 * @author Elias Häußler <elias@haeussler.dev>
 * @license GPL-2.0-or-later
 */
final class Configuration
{
    private const DEFAULT_MODEL = 'gpt-4o-mini';
    private const DEFAULT_TOKENS = 300;
    private const DEFAULT_TEMPERATURE = 0.5;
    private const DEFAULT_NUMBER_OF_COMPLETIONS = 1;
    private const DEFAULT_CACHE_LIFETIME = 60 * 60 * 24; // 1 day
    private const DEFAULT_PROVIDER = ProblemSolving\Solution\Provider\OpenAISolutionProvider::class;
    private const DEFAULT_PROMPT = ProblemSolving\Solution\Prompt\DefaultPrompt::class;

    public function __construct(
        private readonly ConfigurationProvider $provider = new LazyConfigurationProvider(),
    ) {}

    public function getApiKey(): ?string
    {
        $apiKey = $this->provider->get('api/key');

        if (!\is_string($apiKey) || \trim($apiKey) === '') {
            return null;
        }

        return $apiKey;
    }

    public function getModel(): string
    {
        $model = $this->provider->get('attributes/model', self::DEFAULT_MODEL);

        if (!\is_string($model) || \trim($model) === '') {
            $model = self::DEFAULT_MODEL;
        }

        return $model;
    }

    public function getMaxTokens(): int
    {
        $maxTokens = $this->provider->get('attributes/maxTokens', self::DEFAULT_TOKENS);

        if (!\is_numeric($maxTokens) || $maxTokens <= 0) {
            $maxTokens = self::DEFAULT_TOKENS;
        }

        return (int)$maxTokens;
    }

    public function getTemperature(): float
    {
        $temperature = $this->provider->get('attributes/temperature', self::DEFAULT_TEMPERATURE);

        if (!\is_numeric($temperature) || $temperature < 0 || $temperature > 1) {
            $temperature = self::DEFAULT_TEMPERATURE;
        }

        return (float)$temperature;
    }

    public function getNumberOfCompletions(): int
    {
        $numberOfCompletions = $this->provider->get('attributes/numberOfCompletions', self::DEFAULT_NUMBER_OF_COMPLETIONS);

        if (!\is_numeric($numberOfCompletions) || $numberOfCompletions < 1) {
            $numberOfCompletions = self::DEFAULT_NUMBER_OF_COMPLETIONS;
        }

        return (int)$numberOfCompletions;
    }

    public function getCacheLifetime(): int
    {
        $cacheLifetime = $this->provider->get('cache/lifetime', self::DEFAULT_CACHE_LIFETIME);

        if (!\is_numeric($cacheLifetime) || $cacheLifetime < 0) {
            $cacheLifetime = self::DEFAULT_CACHE_LIFETIME;
        }

        return (int)$cacheLifetime;
    }

    public function isCacheEnabled(): bool
    {
        return $this->getCacheLifetime() > 0;
    }

    public function getProvider(): ProblemSolving\Solution\Provider\SolutionProvider
    {
        $providerClass = $this->provider->get('provider', self::DEFAULT_PROVIDER);

        if (!\is_string($providerClass)
            || !\class_exists($providerClass)
            || !\is_a($providerClass, ProblemSolving\Solution\Provider\SolutionProvider::class, true)
        ) {
            $providerClass = self::DEFAULT_PROVIDER;
        }

        return $providerClass::create();
    }

    public function getPrompt(): ProblemSolving\Solution\Prompt\Prompt
    {
        $promptClass = $this->provider->get('prompt', self::DEFAULT_PROMPT);

        if (!\is_string($promptClass)
            || !\class_exists($promptClass)
            || !\is_a($promptClass, ProblemSolving\Solution\Prompt\Prompt::class, true)
        ) {
            $promptClass = self::DEFAULT_PROMPT;
        }

        return $promptClass::create();
    }

    /**
     * @return list<int>
     */
    public function getIgnoredCodes(): array
    {
        $ignoredCodes = $this->provider->get('ignoredCodes', '');

        if (!\is_string($ignoredCodes)) {
            return [];
        }

        return \array_values(
            \array_map(
                intval(...),
                \array_filter(
                    Core\Utility\GeneralUtility::trimExplode(',', $ignoredCodes, true),
                    is_numeric(...),
                ),
            ),
        );
    }
}
